Sun soft shadows

usakhelo opened the pull request:
This PR adds size_multiplier to sun light and soft shadows to sun's sample function.
It also fixes over-brightness when light sample count is less than 1.0
I'll comment code lines for the sake of clarity..

usakhelo commented on src/appleseed/renderer/kernel/lighting/directlightingintegrator.cpp:
```diff
@@ -233,15 +233,19 @@ void DirectLightingIntegrator::compute_outgoing_radiance_light_sampling_low_vari
 
     // Add contributions from non-physical light sources only.
-    for (size_t i = 0, e = m_light_sampler.get_non_physical_light_count(); i < e; ++i)
+    if (m_light_sample_count > 0)
```
I've found that when light sample count < 0 in studio settings direct lights are strangely brighter than usually. In such cases this variable (m_light_sample_count) gets value 0 sometimes and I thought direct light should not be added when it is 0. It fixed the brightness but I'm not sure if fix is correct.

dictoon replied:
> I've found that when light sample count < 0

I assume you mean `light sample count < 1`? Because `light sample count< 0` doesn't make sense and shouldn't be allowed from .studio.

The `light sample count < 1` case is a confirmed bug: https://github.com/appleseedhq/appleseed/issues/1467

I suggest you don't worry about it for now, we'll fix it at some point.

dictoon replied:
This bug was fixed by @Biart95.

usakhelo commented on src/appleseed/renderer/kernel/lighting/directlightingintegrator.cpp:
```diff
@@ -657,7 +662,8 @@ void DirectLightingIntegrator::add_emitting_triangle_sample_contribution(
 
 void DirectLightingIntegrator::add_non_physical_light_sample_contribution(
-    const LightSample&          sample,
+    SamplingContext&            sampling_context,
```
sampling_context added to be passed to sample() function

dictoon replied:
Makes sense.

usakhelo commented on src/appleseed/renderer/kernel/lighting/directlightingintegrator.cpp:
```diff
@@ -670,13 +676,32 @@ void DirectLightingIntegrator::add_non_physical_light_sample_contribution(
+
+    // Generate a uniform sample in [0,1).
+    sampling_context.split_in_place(2, 1);
+    const Vector2d s = sampling_context.next2<Vector2d>();
```
I'm not sure about these two lines. I figured since it is 2d vector dimension should be 2.

dictoon replied:
That's correct, here is the details of how this works:

    sampling_context.split_in_place(2, 1);

This creates a new "trajectory" or "sampling sequence": the 2 means "2D samples" and the 1 means that we want a single sample in that sequence.

    sampling_context.next2<Vector2d>();

This returns the next (and in this case, the only) 2D sample (of type Vector2d, i.e. as two double-precision values). The 2 in `next2()` means that we want a semi-open intervals for samples: [0, 1) (closed on the left, open on the right, i.e. 0 is part of the interval but 1 isn't). That's the only variant available with sampling contexts since it's the only one that we need so far.

With pure random numbers, we have two additional variants: Functions suffixed with 1 such as `rand_double1()` return values in the interval [0, 1] (closed on both ends) and functions suffixed with 3 such as `rand_double3()` return values in the interval (0, 1) (open on both ends).

usakhelo commented on src/appleseed/renderer/kernel/lighting/directlightingintegrator.cpp:
```diff
@@ -670,13 +676,32 @@ void DirectLightingIntegrator::add_non_physical_light_sample_contribution(
+    // Generate a uniform sample in [0,1).
+    sampling_context.split_in_place(2, 1);
+    const Vector2d s = sampling_context.next2<Vector2d>();
+    float probability;
```
this variable is passed to sample to calculate, but I don't know what to do with it later.

dictoon replied:
The principle is the same as for mesh lights:

* Given a light emitting mesh, we first choose a triangle of the mesh with probability `Ptri`. Then, on this triangle, we choose a point with uniform probability `Ppoint` which is simply `1/triangleArea`. The total probability for that point is then `Ptri * Ppoint`.

* Given a set of non-physical lights, we first choose one of the light with probability `Plight` (in this method, `Plight` is `sample.m_probability`). Then, for this light, we choose a direction with probability `Pdir`. For most lights, given a target point to be illuminated, `Pdir` is just 1 since there isn't any choice to be made (given a point light and a target point, there is only one ray joining the two, hence its probability is 1). Similarly as for the mesh light case, the total probability for that light direction is then `Plight * Pdir`.

usakhelo commented on src/appleseed/renderer/kernel/lighting/directlightingintegrator.cpp:
```diff
@@ -670,13 +676,32 @@ void DirectLightingIntegrator::add_non_physical_light_sample_contribution(
+        probability
+    );
+
+    //sample.m_probability *= probability;
```
I've tried to apply calculated probability to light sample but it made light many times brighter.

dictoon replied:
According to my explanation above, that's essentially the correct thing to do. There must be another mistake then, I need to investigate more carefully.

dictoon replied:
Ah, I think I know the problem. The sun's "intensity" we calculate assume that the sun is a pure directional light, i.e. it's the full sun's energy, not the energy per unit of area. So in a sense, _not_ dividing the total contribution by the probability of the sampled direction is correct. To be correct in the general case however, we must do that division (i.e. uncomment the light above, or better: divide by `probability` when computing the overall weight on line 736) and at the same time, we need to divide the sun's intensity by the disk's surface area. These two computations will cancel out, but we need to do it for other light models where they won't necessarily cancel out.

usakhelo commented on src/appleseed/renderer/kernel/lighting/directlightingintegrator.cpp:
```diff
@@ -670,13 +676,32 @@ void DirectLightingIntegrator::add_non_physical_light_sample_contribution(
+    light->sample(
         m_shading_context,
         sample.m_light_transform,
         m_material_sampler.get_point(),
```
target point parameter added to sample()

usakhelo commented on src/appleseed/renderer/kernel/lighting/lighttracing/lighttracingsamplegenerator.cpp:
```diff
@@ -706,6 +706,7 @@ namespace
             light_sample.m_light->sample(
                 m_shading_context,
                 light_sample.m_light_transform,
+                Vector3d(0.0),
```
This and next change calls light's sample() function. I didn't know where to take surface point at this stage so I passed zero vector to make compiler happy.

usakhelo commented on src/appleseed/renderer/kernel/lighting/sppm/sppmphotontracer.cpp:
```diff
@@ -450,6 +450,7 @@ namespace
             light_sample.m_light->sample(
                 shading_context,
                 light_sample.m_light_transform,
+                Vector3d(0.0),
```
same as above

usakhelo commented on src/appleseed/renderer/modeling/light/directionallight.cpp:
```diff
@@ -125,6 +125,7 @@ namespace
         virtual void sample(
             const ShadingContext&   shading_context,
             const Transformd&       light_transform,
+            const Vector3d&         target_point,
```
Dummy target_point parameters, never used in any light except sun light. Third version of sample should added to light.h probably?

dictoon replied:
Yes probably.

usakhelo commented on src/appleseed/renderer/modeling/light/sunlight.cpp:
```diff
@@ -96,6 +96,7 @@ namespace
             m_inputs.declare("environment_edf", InputFormatEntity, "");
             m_inputs.declare("turbidity", InputFormatFloat);
             m_inputs.declare("radiance_multiplier", InputFormatFloat, "1.0");
+            m_inputs.declare("size_multiplier", InputFormatFloat, "1.0");
```
size multiplier parameter added

dictoon replied:
Looks good!

usakhelo commented on src/appleseed/renderer/modeling/light/sunlight.cpp:
```diff
@@ -141,17 +148,24 @@ namespace
+            //angular_diameter = one_radian * sun_diameter / distance
+            //sun_diameter = angular_diameter * distance / one_radian = 0.5331 degrees * 1.496 * 10^8 km / 57.29578 degrees
+            double sun_diameter = 0.5331 * m_safe_scene_diameter / 57.29578;
+            sun_diameter *= m_values.m_size_multiplier;
```
calculating sun's diameter based on distance from scene center

dictoon replied:
Cool, I'll trust your calculations for now :)

usakhelo commented on src/appleseed/renderer/modeling/light/sunlight.cpp:
```diff
@@ -418,6 +437,9 @@ namespace
                 + disk_radius * p[1] * basis.get_tangent_v();
 
+            if (target_point != Vector3d(0.0))
+                outgoing = normalize(target_point - position);
```
this checks for dummy target_point passed from sppmphotontracer.cpp and lighttracingsamplegenerator.cpp

dictoon replied:
That won't work because (0,0,0) is otherwise a perfectrly valid target point. We probably need to find a better mechanism here.

dictoon commented on src/appleseed/renderer/modeling/light/sunlight.cpp:
```diff
@@ -128,6 +129,12 @@ namespace
             m_values.m_turbidity += BaseTurbidity;
 
+            const Source* size_multiplier_src = get_inputs().source("size_multiplier");
+            if (size_multiplier_src && size_multiplier_src->is_uniform())
```
We should probably even emit a warning if the size multiplier is not uniform (uniform = constant) since it wouldn't make much sense.

usakhelo replied:
added warning if size_multiplier is not uniform

dictoon commented on src/appleseed/renderer/modeling/light/sunlight.cpp:
```diff
@@ -141,17 +148,24 @@ namespace
                 s,
                 m_scene_center,
-                m_scene_radius,
+                sun_diameter/2,
```
Would make sense to compute the Sun's _radius_ directly, thus avoiding a division.

dictoon commented on src/appleseed/renderer/modeling/light/sunlight.cpp:
```diff
@@ -486,6 +508,17 @@ DictionaryArray SunLightFactory::get_input_metadata() const
+    metadata.push_back(
+        Dictionary()
+            .insert("name", "size_multiplier")
+            .insert("label", "Sun Size Multiplier")
```
Remove *Sun*.

dictoon commented on src/appleseed/renderer/modeling/light/sunlight.cpp:
```diff
@@ -486,6 +508,17 @@ DictionaryArray SunLightFactory::get_input_metadata() const
+            .insert("max_value", "100.0")
+            .insert("use", "optional")
+            .insert("default", "1.0")
+            .insert("help", "Shadow softness multiplier"));
```
Maybe more explicative: "The size multiplier allows to make the sun bigger or smaller, hence making it cast softer or harder shadows."

usakhelo commented:
Thanks for review!
I'll be fixing this pr according to your comments. Should I close it for now or is it ok if it will be updated from time to time? I'll do rebase when it'll be final.

P.S. I have to say that lighting is very different when `sample()` takes `target_point` into account. Without it lighting is flat. I don't have good example but here's what I have (upper is with target_point):
![image](https://user-images.githubusercontent.com/1850877/27671731-0376f92e-5c64-11e7-845c-9498fd2b3317.png)

dictoon commented:
Hey, thanks! I'll check your changes when I have a moment.

> Should I close it for now or is it ok if it will be updated from time to time? I'll do rebase when it'll be final.

It's actually best to keep it open and amend it until it reaches a point where we can merge it, and then rebase it.

> I have to say that lighting is very different when sample() takes target_point into account

That's kind of surprising, since the sun is so far away (in theory) the lighting should be rather flat, like in your second picture. What's the sun size in this case?

dictoon commented on src/appleseed/renderer/kernel/lighting/directlightingintegrator.cpp:
```diff
@@ -670,13 +676,32 @@ void DirectLightingIntegrator::add_non_physical_light_sample_contribution(
+
+    //sample.m_probability *= probability;
+
+    //light->evaluate(
```
Remove dead code.

dictoon commented on src/appleseed/renderer/modeling/light/light.h:
```diff
@@ -107,6 +107,15 @@ class APPLESEED_DLLSYMBOL Light
+        foundation::Vector3d&           position,                   // world space emission position
+        foundation::Vector3d&           outgoing,                   // world space emission direction, unit-length
+        Spectrum&                       value,                      // light value
+        float&                          probability) const;     // PDF value
```
Align comments.

usakhelo commented:
Hi Franz,
I've fixed some of the things:

* Probability applied to all the lights. In case of sunlight probability = 1/sun_area, so it is applied to light value and then `weight` is divided by it to compensate. For directional light I've made `probability =1.0` and `disk_radius=0`
* overloaded `sample()` function with `target_point` argument added to all the lights, so they look as expected. 
* added separate `sample_disk_parallel()` function to sunlight to be called from `sample()` without `target_point` argument
* removed `DirectLightingIntegrator` fixes

Also you asked me about sun size on those renderings - sun diameter is 0.74mm and distance is 88mm (`m_scene_radius`)

dictoon commented on src/appleseed/renderer/kernel/lighting/directlightingintegrator.cpp:
```diff
@@ -708,7 +720,7 @@ void DirectLightingIntegrator::add_non_physical_light_sample_contribution(
     const float attenuation = light->compute_distance_attenuation(
         m_material_sampler.get_point(), emission_position);
-    const float weight = transmission * attenuation / sample.m_probability;
+    const float weight = transmission * attenuation / sample.m_probability / probability;
```
You can avoid one division by dividing by the product of probabilities:

    const float weight = transmission * attenuation / (sample.m_probability * probability);

This is the kind of optimization the compiler probably cannot do because it may slightly change the result (unless "fast math" is enabled, which it isn't in appleseed).

dictoon commented on src/appleseed/renderer/kernel/lighting/directlightingintegrator.h:
```diff
@@ -144,7 +144,8 @@ class DirectLightingIntegrator
     void add_non_physical_light_sample_contribution(
-        const LightSample&              sample,
+        SamplingContext&                sampling_context,
+        LightSample&                    sample,
```
Why is `sample` passed as a non-const reference?

dictoon commented on src/appleseed/renderer/modeling/light/directionallight.cpp:
```diff
@@ -242,7 +242,7 @@ namespace
             value = m_values.m_irradiance;
 
-            probability = 1.0f / (Pi<float>() * square(static_cast<float>(disk_radius)));
+            probability = 1.0f;
```
That's only correct when `disk_radius` is 0.

You shouldn't change the way the probability is computed, it was correct before. What you need to do instead is no longer call `sample_disk()` when you have a target point. The code will also be much simpler in that case.

dictoon commented on src/appleseed/renderer/modeling/light/directionallight.cpp:
```diff
@@ -135,7 +135,7 @@ namespace
                 s,
                 m_scene_center,
-                m_scene_radius,
+                0,
```
Here, since you have a target point, you can compute `position`, `outgoing` and `probability` (and `value` but this one doesn't change) in a much simpler manner.

dictoon commented on src/appleseed/renderer/modeling/light/maxomnilight.cpp:
```diff
@@ -125,6 +142,7 @@ namespace
             outgoing = sample_sphere_uniform(s);
             value = m_values.m_intensity;
             probability = RcpFourPi<float>();
+            value *= probability;
```
`value` is not supposed to be multiplied by probability here, it's the reason why we return both of them separately.

dictoon commented on src/appleseed/renderer/modeling/light/maxomnilight.cpp:
```diff
@@ -115,6 +115,23 @@ namespace
+            position = light_transform.get_parent_origin();
+            outgoing = normalize(target_point - position);
+            value = m_values.m_intensity;
+            probability = RcpFourPi<float>();
```
Since there is no sampling, I believe that in this case `probability` should be 1.

dictoon commented on src/appleseed/renderer/modeling/light/maxspotlight.cpp:
```diff
@@ -139,6 +140,31 @@ namespace
             compute_radiance(shading_context, light_transform, axis, outgoing, value);
+
+            outgoing = normalize(target_point - position);
+            value *= probability;
```
Don't multiply value by probability.

usakhelo commented:
Hi,
Pushed another attempt to fix comments.
I moved `light_value *= probability` to DirectLightingIntegrator, is that right?

dictoon commented:
Superseded by #1519.
